mycli/output_formatter.py

```diff
@@ -0,0 +1,113 @@
+
+from __future__ import unicode_literals
+
+import csv
+import binascii
+try:
+    from cStringIO import StringIO
+except ImportError:
+    from io import StringIO
+
+from tabulate import tabulate
+
+from .packages.expanded import expanded_table
+
+
+def override_missing_value(data, missing_value='', **_):
+    """Override missing values in the data with *missing_value*."""
+    return [[missing_value if v is None else v for v in row] for row in data]
+
+
+def bytes_to_unicode(data, **_):
+    results = []
+    for row in data:
+        result = []
+        for v in row:
+            if isinstance(v, bytes):
+                try:
+                    conv = v.decode('utf8')
+                except:
+                    v = '0x' + binascii.hexlify(v).decode('ascii')
+            result.append(v)
+        results.append(result)
+    return results
+
+
+def tabulate_wrapper(data, headers, table_format=None, missing_value='', **_):
+    """Wrap tabulate inside a standard function for OutputFormatter."""
+    return tabulate(data, headers, tablefmt=table_format,
+                    missingval=missing_value, disable_numparse=True)
+
+
+def csv_wrapper(data, headers, delimiter=',', **_):
+    """Wrap CSV formatting inside a standard function for OutputFormatter."""
+    content = StringIO()
+    writer = csv.writer(content, delimiter=str(delimiter))
+    writer.writerow(headers)
+
+    for row in data:
+        writer.writerow(row)
+
+    output = content.getvalue()
+    content.close()
+
+    return output
+
+
+class OutputFormatter(object):
+    """A class with a standard interface for various formatting libraries."""
+
+    def __init__(self):
+        """Register the supported output formats."""
+        self._output_formats = {}
+
+        tabulate_formats = ('plain', 'simple', 'grid', 'fancy_grid', 'pipe',
+                            'orgtbl', 'jira', 'psql', 'rst', 'mediawiki',
+                            'moinmoin', 'html', 'latex', 'latex_booktabs',
+                            'textile')
+        for tabulate_format in tabulate_formats:
+            self.register_output_format(tabulate_format, tabulate_wrapper,
+                                        preprocessor=bytes_to_unicode,
+                                        table_format=tabulate_format,
+                                        missing_value='<null>')
+
+        self.register_output_format('csv', csv_wrapper,
+                                    preprocessor=override_missing_value,
+                                    missing_value='null')
+        self.register_output_format('tsv', csv_wrapper, delimiter='\t',
+                                    preprocessor=override_missing_value,
+                                    missing_value='null')
+
+        self.register_output_format('expanded', expanded_table,
+                                    preprocessor=override_missing_value,
+                                    missing_value='<null>')
+
```

Comment:
Hmm yes, that's true. Let's keep it this way. I was thinking of letting the user configure more output formatters at runtime but we can always refactor it by then (now it's too far fetched).